src/Autofac/Diagnostics/DiagnosticTracerBase.cs

```diff
@@ -0,0 +1,367 @@
+        /// names are case-sensitive.
+        /// </param>
+        /// <remarks>
+        /// <para>
+        /// By default the observer is not subscribed to any events. You must
+        /// opt-in to any events you wish to handle. You may use
+        /// <see cref="EnableAll"/> to subscribe to all Autofac events at once.
+        /// </para>
+        /// </remarks>
+        /// <seealso cref="EnableAll"/>
+        /// <seealso cref="Disable"/>
+        public void Enable(string diagnosticName)
+        {
+            if (diagnosticName == null)
+            {
+                throw new ArgumentNullException(nameof(diagnosticName));
+            }
+
+            if (!_subscriptions.Contains(diagnosticName))
+            {
+                _subscriptions.Add(diagnosticName);
+            }
+        }
+
+        /// <summary>
+        /// Subscribes the observer to all Autofac events.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// By default the observer is not subscribed to any events. You must
+        /// opt-in to any events you wish to handle. This method helps you
+        /// to quickly subscribe to all the Autofac events at once. You may use
+        /// <see cref="Enable"/> to subscribe to individual events.
+        /// </para>
+        /// </remarks>
+        /// <seealso cref="Enable"/>
+        /// <seealso cref="Disable"/>
+        public void EnableAll()
+        {
+            _subscriptions.Add(DiagnosticEventKeys.MiddlewareStart);
+            _subscriptions.Add(DiagnosticEventKeys.MiddlewareFailure);
+            _subscriptions.Add(DiagnosticEventKeys.MiddlewareSuccess);
+            _subscriptions.Add(DiagnosticEventKeys.OperationFailure);
+            _subscriptions.Add(DiagnosticEventKeys.OperationStart);
+            _subscriptions.Add(DiagnosticEventKeys.OperationSuccess);
+            _subscriptions.Add(DiagnosticEventKeys.RequestFailure);
+            _subscriptions.Add(DiagnosticEventKeys.RequestStart);
+            _subscriptions.Add(DiagnosticEventKeys.RequestSuccess);
+        }
+
+        /// <summary>
+        /// Unsubscribes the observer from a particular named diagnostic event.
+        /// </summary>
+        /// <param name="diagnosticName">
+        /// The name of the event to which the observer should unsubscribe. Diagnostic
+        /// names are case-sensitive.
+        /// </param>
+        /// <remarks>
+        /// <para>
+        /// By default the observer is not subscribed to any events. You must
+        /// opt-in to any events you wish to handle. You may use
+        /// <see cref="EnableAll"/> to subscribe to all Autofac events at once,
+        /// or <see cref="Enable"/> to subscribe to individual events.
+        /// </para>
+        /// </remarks>
+        /// <seealso cref="EnableAll"/>
+        /// <seealso cref="Enable"/>
+        public void Disable(string diagnosticName)
```

Comment:
I had them virtual, but you shouldn't call virtual methods from a constructor due to issues with order of constructor execution and how it could lead to calling the method in a class not fully constructed.

I could separate Enable and Disable into public virtual and protected non-virtual methods such that the ctor and default public methods call the protected methods.

Reply:
I split these into `public virtual Enable` and `protected EnableBase` such that constructors and things can call the non-virtual versions but we can override them to "disable" subscription management. I did that on the default tracer - it now throws a `NotSupportedException` if you try to `Enable` or `Disable` a subscription.